crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -263,6 +263,11 @@ use std::{cell::UnsafeCell, marker::PhantomData};
 pub unsafe trait QueryData: WorldQuery {
     /// The read-only variant of this [`QueryData`], which satisfies the [`ReadOnlyQueryData`] trait.
     type ReadOnly: ReadOnlyQueryData<State = <Self as WorldQuery>::State>;
+    /// The "reffed" variant of this [`QueryData`] (as in, the [`Ref`] smart pointer for change detection).
+    /// - If this [`QueryData`] doesn't have a "reffed" form (for example, [`Entity`]) - this type must be identical to [`Self::ReadOnly`].
+    /// - If this [`QueryData`] does have a "reffed" form (for example, &'w T -> Ref<'w, T>) then that would be this type.
```

Comment:
I agree, the "missing API" (such as `iter_ref`, `get_single_ref`, etc.) should be completed, either here or in another PR.
I don't know if these should be added to the safety spec because the current one is specific enough.